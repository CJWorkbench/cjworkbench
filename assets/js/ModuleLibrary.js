import PropTypes from 'prop-types';
import React from 'react';
import ModuleCategory from './ModuleCategory';
import ImportModuleFromGitHub from './ImportModuleFromGitHub';
import Module from './Module';
import ModuleSearch from './ModuleSearch';


/**
 * Contains the Module Library. The Module Library is effectively the place
 * that allows users to browse all the modules in the system that are
 * available to them. Therefore, the server should only send through the
 * modules the users are entitled to, and this class will render those modules
 * accordingly, and allow users to add them to workflows.
 *
 * Note: the current implementation mandates that this library appears vis-à-vis
 * the workflow, and is not an independent component. Perhaps, at some point, it
 * should be, so that newcomers to the system can get an idea as to the modules
 * that are supported – both, those created by us and those created by third
 * parties.
 */
export default class ModuleLibrary extends React.Component {
  constructor(props) {
    super(props);

    // Do we have any modules at all? If not, "Add Data" category starts open
    var workflowEmpty = (!props.workflow.wf_modules || !props.workflow.wf_modules.length);

    this.state = {
      libraryOpen: !this.props.isReadOnly,
      openCategory: workflowEmpty ? "Add data" : null,
      items: [],
    };
    this.addModule = this.props.addModule.bind(this);
    this.setOpenCategory = this.setOpenCategory.bind(this);
    this.toggleLibrary = this.toggleLibrary.bind(this);
    this.updated = this.updated.bind(this);
  }

  /**
   * Queries server for all the available modules for the given credentials.
   * The response that's returned should be a list, where each item has the
   * following properties:
   * - name (e.g. "Paste CSV")
   * - id (e.g. 8)
   * - category (e.g. "Sources")
   * - description (e.g. "Allows users to copy in a CSV from an external source.")
   *
   * Additionally, these properties are optional (and mostly used by modules
   * imported from GitHub, i.e. not the core modules):
   * - icon
   * - source
   * - author
   * - version
   */
  componentWillMount() {
    if (!this.props.isReadOnly) { // don't load modules if we can't open library
      this.props.api.getModules()
        .then(json => {
          // Sort modules – first by category, then by name
          json.sort((a, b) => {
            if (a.category > b.category) {
              return 1;
            } else if (a.category < b.category) {
              return -1;
            } else if (a.name > b.name) {
              return 1;
            } else if (a.name < b.name) {
              return -1;
            } else {
              return 0;
            }
          });
          this.setState({items: json});
        })
    }
  }

  // Categories call this to indicate that they've been opened, so we can close all the rest
  setOpenCategory(name) {
      this.setState({openCategory: name});
  }

  itemClick(event) {
    this.props.addModule(event.target.id);
  }

  updated() {
    this.componentWillMount() // dummy update to force a re-render.
  }

  toggleLibrary() {
    if (!this.props.isReadOnly) {
      this.setState({ libraryOpen: !this.state.libraryOpen });
    }
  }

  // Return an array of <Module Category>, each of which has child <Module>s.
  renderCategories() {
    // This assumes that the items are already sorted by category,
    // which happens in {code: componentDidMount}. So, if someone
    // changes that, there is a good chance that this will result in
    // unexpected behaviour.
    var modules = this.state.items;
    var currentCategory = null;
    var modulesByCategory = [];
    var categories = [];

    for (var item of modules) { // Yes, for...of is ES6 syntax, and yes, it's gross.

      let module = <Module
        key={item.name}
        name={item.name}
        icon={item.icon}
        id={item.id}
        addModule={this.props.addModule}
      />;

      if (currentCategory  === null) {
        currentCategory  = item.category;
      } else if (currentCategory !== item.category) {
        // We should only create the ModuleCategory once we have all modules for given category.

        // console.log("Creating category " +  currentCategory);

        // Start Add Data open if there is nothing in the Workflow
        let moduleCategory = <ModuleCategory
          name={currentCategory }
          key={currentCategory }
          modules={modulesByCategory}
          isReadOnly={this.props.isReadOnly}
          collapsed={currentCategory != this.state.openCategory}
          setOpenCategory={this.setOpenCategory}
        />;
        categories.push(moduleCategory);
        modulesByCategory = [];
        currentCategory  = item.category;
      }
      modulesByCategory.push(module);
    }

    // the last item / category
    if (currentCategory  != null) {  // modules may not be loaded yet
      // console.log("Creating final category " +  currentCategory);

      let moduleCategory = <ModuleCategory
        name={currentCategory }
        key={currentCategory }
        modules={modulesByCategory}
        isReadOnly={this.props.isReadOnly}
        collapsed={currentCategory != this.state.openCategory}
        setOpenCategory={this.setOpenCategory}
      />;
      categories.push(moduleCategory);
    }

    return categories;
  }

  // Main render.
  render() {

    // console.log("render...");

    if (this.state.libraryOpen) {
      // Outermost div seems necessary to set background color below ImportFromGithub
      return (
        <div>
          <div className='module-library-open'>
            <div className='library-nav-bar'>

              <div className='d-flex align-items-center flex-row mb-4'>
                <a href="/workflows" className="logo"><img src="/static/images/logo.png" width="20"/></a>
                <a href="/workflows" className='logo-2 ml-3 t-vl-gray '>Workbench</a>
                <div className='icon-sort-left-vl-gray ml-auto mt-2 close-open-toggle' onClick={this.toggleLibrary}></div>
              </div>

              <ModuleSearch addModule={this.props.addModule}
                            items={this.state.items}
                            workflow={this.props.workflow}/>
            </div>

            <div className="list">
              {this.renderCategories()}
            </div>

            <ImportModuleFromGitHub moduleAdded={this.updated}/>
          </div>
        </div>
      )
    } else {
      return (
        <div className='module-library-collapsed'>
          <div className="expand-lib">
            <div className="expand-lib-button d-flex">
              <div className="logo" onClick={this.toggleLibrary}><img src="/static/images/logo.png" width="20"/></div>
              {
                (this.props.isReadOnly)
                  ? null
                  : <div
                      className='icon-sort-right-vl-gray ml-auto ml-3 mt-2 close-open-toggle'
                      onClick={this.toggleLibrary}>
                    </div>
              }
            </div>
          </div>
        </div>
      )
    }
  }

}

ModuleLibrary.propTypes = {
  addModule: PropTypes.func.isRequired,
  workflow:  PropTypes.object.isRequired,
  api:       PropTypes.object.isRequired,
  isReadOnly: PropTypes.bool.isRequired,
};
