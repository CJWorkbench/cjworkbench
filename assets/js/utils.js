// ---- Utilities ---
import * as Cookies from 'js-cookie'
import { fromByteArray as base64Encode } from 'base64-js'
import { t } from '@lingui/macro'

export function goToUrl (url) {
  window.location.href = url
}

// Current CSRF token
export const csrfToken = Cookies.get('csrftoken')

// Gets the letter coordinate of a column from its index within the column names array
export function idxToLetter (idx) {
  var letters = ''
  var cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode(cidx % 26 + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}

// Log to Intercom, if installed
export function logUserEvent (name, metadata) {
  if (!window.APP_ID) return

  // If we're in a lesson, drop the event.
  if (window.initState && window.initState.lessonData) return

  window.Intercom('trackEvent', name, metadata)
}

export function timeDifference (start, end, i18n) {
  const ms = new Date(end) - new Date(start)
  const seconds = Math.floor(ms / 1000)
  const minutes = Math.floor(seconds / 60)
  const hours = Math.floor(minutes / 60)
  const days = Math.floor(hours / 24)
  const years = Math.floor(days / 365.25)

  if (years > 0) {
    return i18n._(t('js.util.timeDifference.ago.years')`${years}y ago`)
  } else if (days > 0) {
    return i18n._(t('js.util.timeDifference.ago.days')`${days}d ago`)
  } else if (hours > 0) {
    return i18n._(t('js.util.timeDifference.ago.hours')`${hours}h ago`)
  } else if (minutes > 0) {
    return i18n._(t('js.util.timeDifference.ago.minutes')`${minutes}m ago`)
  } else if (seconds > 0) {
    return i18n._(t('js.util.timeDifference.ago.seconds')`${seconds}s ago`)
  } else {
    return i18n._(t('js.util.timeDifference.now')`just now`)
  }
}

export function escapeHtml (str) {
  str
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;')

  return str
}

/**
 * Scroll `containerEl` vertically by the smallest amount possible to put
 * `el` in view.
 *
 * Set the container CSS `scroll-behavior: smooth` to make this transition
 * look nice.
 *
 * @param el Element we want to focus
 * @param containerEl Element we'll set scrollTop on so it contains el
 * @param marginTop Minimum number of pixels between containerEl.top and el.top
 * @param marginBottom Minimum number of pixels between containerEl.bottom and el.bottom
 */
export function scrollTo (el, containerEl, marginTop, marginBottom) {
  if (marginTop === undefined) marginTop = 10
  if (marginBottom === undefined) marginBottom = 10

  const elRect = el.getBoundingClientRect()
  const containerRect = containerEl.getBoundingClientRect()

  // Calculate dy: how much do we need to add to containerEl.scrollTop to
  // make el go where we want it to go?
  let dy = 0 // Prefer not to scroll
  if (elRect.bottom + marginBottom > containerRect.bottom) {
    // Scroll down if we need to.
    // We do this before up because some els might be taller than containerEl,
    // and in that case we want scroll-up to override scroll-down (because the
    // top of el is more important than the bottom).
    dy += (elRect.bottom + marginBottom - containerRect.bottom)
  }
  if (dy + elRect.top - marginTop < containerRect.top) {
    // Scroll up if we need to.
    dy -= (containerRect.top - (elRect.top - marginTop))
  }

  containerEl.scrollTop += dy
}

/**
 * Build a unique slug with the given prefix.
 *
 * The slug uses randomly-generated bytes: 9 bytes, base64-encoded with
 * '+' and '/' changed to '-' and '_' (to fit a Django "slug" field).
 *
 * Django 'slug' regex: r'^[-a-zA-Z0-9_]+\Z'
 *
 * According to https://en.wikipedia.org/wiki/Birthday_attack, 9 bytes -- 72
 * bits -- means if clients call this 97,000 times, they have a <10^-12 chance
 * of generating a collision within a given Workflow. This should protect even
 * the most click-happy user. (6 bytes would allow 380 before the chance grows
 * higher, and 380 steps is possible.)
 *
 * (Why 9 bytes and not 8? Because base64-encoding gives the same string
 * length. 9 bytes encoded as base64 gives 12 bytes.)
 */
export function generateSlug (prefix) {
  const bytes = new Uint8Array(9)
  window.crypto.getRandomValues(bytes)
  return prefix + (
    base64Encode(bytes)
      .replace(/\+/g, '-')
      .replace(new RegExp('/', 'g'), '_')
  )
}
