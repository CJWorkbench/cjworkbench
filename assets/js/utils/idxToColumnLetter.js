/**
 * Return 'A' for 0, 'B' for 1, 'AA' for 26, and so on.
 */
export default function idxToColumnLetter (idx) {
  let letters = ''
  let cidx = parseInt(idx)
  cidx += 1
  do {
    cidx -= 1
    letters = String.fromCharCode((cidx % 26) + 65) + letters
    cidx = Math.floor(cidx / 26)
  } while (cidx > 0)
  return letters
}
