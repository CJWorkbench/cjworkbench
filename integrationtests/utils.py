# Utilities for integration tests

import subprocess
import unittest
import email.message
from typing import Optional
import re

from integrationtests.browser import Browser
from integrationtests.helpers import accounts


_url_regex = re.compile('https?://[^\\s]+')

def find_url_in_email(message: email.message.Message) -> Optional[str]:
    """Return the first URL in the given message's payload, or None."""
    body = message.get_payload()
    match = _url_regex.search(body)

    if not match: return None
    return match.group(0)


def _find_server_url():
    """Get URL using `docker port`"""
    process = subprocess.run([
        'docker',
        'port',
        'cjworkbench_integrationtest_django',
        '8000/tcp'
    ], stdout=subprocess.PIPE)
    port_str = process.stdout.decode('ascii').split(':')[1].strip()
    return f"http://localhost:{port_str}"


class WorkbenchBase(unittest.TestCase):
    serve_static = True
    account_admin = accounts.AccountAdmin()
    live_server_url = _find_server_url()

    def setUp(self):
        super().setUp()

        self.account_admin.clear_data_from_previous_tests()

        #self.current_site = Site.objects.get_current()
        #self.SocialApp1 = self.current_site.socialapp_set.create(
        #    provider="facebook",
        #    name="Facebook",
        #    client_id="1234567890",
        #    secret="0987654321",
        #)
        #self.SocialApp2 = self.current_site.socialapp_set.create(
        #    provider="google",
        #    name="Google",
        #    client_id="1234567890",
        #    secret="0987654321",
        #)

        self.browser = Browser(base_url=self.live_server_url)


    def tearDown(self):
        self.browser.quit()


# Derive from this to perform all tests logged in
class LoggedInIntegrationTest(WorkbenchBase):
    def setUp(self):
        super().setUp()

        self.user = self.account_admin.create_user('user@example.org')
        self.user_email = self.account_admin.verify_user_email(self.user)

        accounts.login(self.browser, self.user.email, self.user.email)


    def tearDown(self):
        self.account_admin.destroy_user_email(self.user_email)
        self.account_admin.destroy_user(self.user)
        self.account_admin.destroy_modules()

        super().tearDown()
