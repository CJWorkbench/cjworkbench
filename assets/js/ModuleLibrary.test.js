/**
 * Testing Stories:
 * -In not-read-only, renders <ModuleLibraryOpen> by default in not-read-only 
 * -When read-only, renders <ModuleLibraryClosed> without modules.
 * 
 */

import React from 'react'
import ModuleLibrary  from './ModuleLibrary'
import { mount } from 'enzyme'
import { jsonResponseMock, emptyAPI } from './utils'
import HTML5Backend from 'react-dnd-html5-backend'
import { DragDropContextProvider } from 'react-dnd'


var wrapper;
var addModule =  () => {};
var dropModule =  () => {};
var workflow = {
  "id":15,
  "name":"What a workflow!",
};
var modules = [
  {
    "id":1,
    "name":"Chartbuilder",
    "category":"Visualize",
    "description":"Create line, column and scatter plot charts.",
    "icon":"chart"
  },
  {
    "id":2,
    "name":"Load from Facebork",
    "category":"Add data",
    "description":"Import from your favorite snowshall media",
    "icon":"url"
  },
  {
    "id":4,
    "name":"Load from Enigma",
    "category":"Add data",
    "description":"Connect a dataset from Enigma's collection via URL.",
    "icon":"url"
  }
];
var api = {};

describe('ModuleLibrary', () => {

  describe('Not Read-only', () => {
  
    beforeEach(() => {
      api = {
        getModules: jsonResponseMock(modules),
        setWfLibraryCollapse: jest.fn()
      };
      wrapper = mount(
        <DragDropContextProvider backend={HTML5Backend}>
          <ModuleLibrary
            addModule={addModule}
            dropModule={dropModule}
            api={api}
            workflow={workflow}
            isReadOnly={false}
          />
        </DragDropContextProvider>);
    });
    afterEach(() => wrapper.unmount());  
    
    it('Renders in open state and loads modules', (done) => {
      expect(wrapper).toMatchSnapshot();
      // should have called API for its data on componentDidMount
      expect(api.getModules.mock.calls.length).toBe(1);
      // check that Library is open
      expect(wrapper.find('.module-library-open')).toHaveLength(1);
      // let json promise resolve (wait for modules to load)
      setImmediate( () => {
        expect(wrapper).toMatchSnapshot();
        // check that module categories have loaded
        expect(wrapper.find('.ml-cat')).toHaveLength(2);
        // check that modules have loaded
        expect(wrapper.find('.ml-list .ml-icon-container')).toHaveLength(3);
        done();
      });
    });

    it('Clicking on arrow will invoke API to toggle collapse', (done) => {
      // let json promise resolve (wait for modules to load)
      setImmediate( () => {
        let arrow = wrapper.find('.close-open-toggle');
        expect(arrow).toHaveLength(1);
        arrow.simulate('click');
        expect(api.setWfLibraryCollapse.mock.calls.length).toBe(1);
        done();
      });
    });

  });

  describe('Read-only', () => {
    
    beforeEach(() => {
      api = {
        getModules: jsonResponseMock(modules),
        setWfLibraryCollapse: jest.fn()
      };
      wrapper = mount(
        <DragDropContextProvider backend={HTML5Backend}>
          <ModuleLibrary
            addModule={addModule}
            dropModule={dropModule}
            api={api}
            workflow={workflow}
            isReadOnly={true}
          />
        </DragDropContextProvider>);
    });
    afterEach(() => wrapper.unmount());  
    
    it('Renders in closed state, without modules', () => {
      expect(wrapper).toMatchSnapshot();
      // should NOT call getModules 
      expect(api.getModules.mock.calls.length).toBe(0);
      // check that Library is closed
      expect(wrapper.find('.module-library-closed')).toHaveLength(1);
    });

  });

});


