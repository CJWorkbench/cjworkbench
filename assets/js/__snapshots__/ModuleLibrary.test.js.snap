// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleLibrary renders closed when read-only 1`] = `
<ModuleLibrary
  addModule={[Function]}
  api={Object {}}
  isReadOnly={true}
  workflow={
    Object {
      "id": 15,
      "name": "What a workflow!",
    }
  }
>
  <div
    className="module-library-collapsed"
  >
    <div
      className="expand-lib"
    >
      <div
        className="expand-lib-button d-flex"
      >
        <div
          className="logo"
          onClick={[Function]}
        >
          <img
            src="/static/images/logo.png"
            width="20"
          />
        </div>
      </div>
    </div>
  </div>
</ModuleLibrary>
`;

exports[`ModuleLibrary renders open when not read-only, with list of module categories 1`] = `
<ModuleLibrary
  addModule={[Function]}
  api={
    Object {
      "getModules": [Function],
    }
  }
  isReadOnly={false}
  workflow={
    Object {
      "id": 15,
      "name": "What a workflow!",
    }
  }
>
  <div>
    <div
      className="module-library-open"
    >
      <div
        className="library-nav-bar"
      >
        <div
          className="d-flex align-items-center flex-row mb-4"
        >
          <a
            className="logo"
            href="/workflows"
          >
            <img
              src="/static/images/logo.png"
              width="20"
            />
          </a>
          <a
            className="logo-2 ml-3 t-vl-gray "
            href="/workflows"
          >
            Workbench
          </a>
          <div
            className="icon-sort-left-vl-gray ml-auto mt-2 close-open-toggle"
            onClick={[Function]}
          />
        </div>
        <ModuleSearch
          addModule={[Function]}
          items={Array []}
          workflow={
            Object {
              "id": 15,
              "name": "What a workflow!",
            }
          }
        >
          <Autosuggest
            alwaysRenderSuggestions={false}
            focusInputOnSuggestionClick={true}
            getSectionSuggestions={[Function]}
            getSuggestionValue={[Function]}
            highlightFirstSuggestion={false}
            id="1"
            inputProps={
              Object {
                "onChange": [Function],
                "placeholder": "Search Modules",
                "value": "",
              }
            }
            multiSection={true}
            onSuggestionSelected={[Function]}
            onSuggestionsClearRequested={[Function]}
            onSuggestionsFetchRequested={[Function]}
            renderSectionTitle={[Function]}
            renderSuggestion={[Function]}
            renderSuggestionsContainer={[Function]}
            shouldRenderSuggestions={[Function]}
            suggestions={Array []}
            theme={
              Object {
                "container": "react-autosuggest__container",
                "containerOpen": "react-autosuggest__container--open",
                "input": "react-autosuggest__input",
                "inputFocused": "react-autosuggest__input--focused",
                "inputOpen": "react-autosuggest__input--open",
                "sectionContainer": "react-autosuggest__section-container",
                "sectionContainerFirst": "react-autosuggest__section-container--first",
                "sectionTitle": "react-autosuggest__section-title",
                "suggestion": "react-autosuggest__suggestion",
                "suggestionFirst": "react-autosuggest__suggestion--first",
                "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
                "suggestionsContainer": "react-autosuggest__suggestions-container",
                "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
                "suggestionsList": "react-autosuggest__suggestions-list",
              }
            }
          >
            <Autowhatever
              getSectionItems={[Function]}
              highlightedItemIndex={null}
              highlightedSectionIndex={null}
              id="1"
              inputProps={
                Object {
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onFocus": [Function],
                  "onKeyDown": [Function],
                  "placeholder": "Search Modules",
                  "value": "",
                }
              }
              itemProps={[Function]}
              items={Array []}
              multiSection={true}
              renderInputComponent={[Function]}
              renderItem={[Function]}
              renderItemData={
                Object {
                  "query": "",
                }
              }
              renderItemsContainer={[Function]}
              renderSectionTitle={[Function]}
              theme={
                Object {
                  "container": "react-autosuggest__container",
                  "containerOpen": "react-autosuggest__container--open",
                  "input": "react-autosuggest__input",
                  "inputFocused": "react-autosuggest__input--focused",
                  "inputOpen": "react-autosuggest__input--open",
                  "item": "react-autosuggest__suggestion",
                  "itemFirst": "react-autosuggest__suggestion--first",
                  "itemHighlighted": "react-autosuggest__suggestion--highlighted",
                  "itemsContainer": "react-autosuggest__suggestions-container",
                  "itemsContainerOpen": "react-autosuggest__suggestions-container--open",
                  "itemsList": "react-autosuggest__suggestions-list",
                  "sectionContainer": "react-autosuggest__section-container",
                  "sectionContainerFirst": "react-autosuggest__section-container--first",
                  "sectionTitle": "react-autosuggest__section-title",
                }
              }
            >
              <div
                className="react-autosuggest__container"
              >
                <input
                  aria-activedescendant={null}
                  aria-autocomplete="list"
                  aria-expanded={false}
                  aria-haspopup={false}
                  aria-owns="react-autowhatever-1"
                  autoComplete="off"
                  className="react-autosuggest__input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Search Modules"
                  role="combobox"
                  type="text"
                  value=""
                />
                <div
                  className="react-autosuggest__suggestions-container"
                  id="react-autowhatever-1"
                />
              </div>
            </Autowhatever>
          </Autosuggest>
        </ModuleSearch>
      </div>
      <div
        className="list"
      />
      <ImportModuleFromGitHub
        moduleAdded={[Function]}
      >
        <div>
          <div
            className="import-module-button content-3 mb-5 t-vl-gray"
            onClick={[Function]}
          >
            IMPORT FROM GITHUB
          </div>
          ;
          <Modal
            autoFocus={true}
            backdrop={true}
            backdropTransitionTimeout={150}
            className="dialog-window"
            fade={true}
            isOpen={false}
            keyboard={true}
            modalTransitionTimeout={300}
            toggle={[Function]}
            zIndex={1050}
          />
        </div>
      </ImportModuleFromGitHub>
    </div>
  </div>
</ModuleLibrary>
`;

exports[`ModuleLibrary renders open when not read-only, with list of module categories 2`] = `
<ModuleLibrary
  addModule={[Function]}
  api={
    Object {
      "getModules": [Function],
    }
  }
  isReadOnly={false}
  workflow={
    Object {
      "id": 15,
      "name": "What a workflow!",
    }
  }
>
  <div>
    <div
      className="module-library-open"
    >
      <div
        className="library-nav-bar"
      >
        <div
          className="d-flex align-items-center flex-row mb-4"
        >
          <a
            className="logo"
            href="/workflows"
          >
            <img
              src="/static/images/logo.png"
              width="20"
            />
          </a>
          <a
            className="logo-2 ml-3 t-vl-gray "
            href="/workflows"
          >
            Workbench
          </a>
          <div
            className="icon-sort-left-vl-gray ml-auto mt-2 close-open-toggle"
            onClick={[Function]}
          />
        </div>
        <ModuleSearch
          addModule={[Function]}
          items={
            Array [
              Object {
                "category": "Add data",
                "description": "Connect a dataset from Enigma's collection via URL.",
                "icon": "url",
                "id": 4,
                "name": "Load from Enigma",
              },
              Object {
                "category": "Add data",
                "description": "Import from your favorite snowshall media",
                "icon": "url",
                "id": 2,
                "name": "Load from Facebork",
              },
              Object {
                "category": "Visualize",
                "description": "Create line, column and scatter plot charts.",
                "icon": "chart",
                "id": 1,
                "name": "Chartbuilder",
              },
            ]
          }
          workflow={
            Object {
              "id": 15,
              "name": "What a workflow!",
            }
          }
        >
          <Autosuggest
            alwaysRenderSuggestions={false}
            focusInputOnSuggestionClick={true}
            getSectionSuggestions={[Function]}
            getSuggestionValue={[Function]}
            highlightFirstSuggestion={false}
            id="1"
            inputProps={
              Object {
                "onChange": [Function],
                "placeholder": "Search Modules",
                "value": "",
              }
            }
            multiSection={true}
            onSuggestionSelected={[Function]}
            onSuggestionsClearRequested={[Function]}
            onSuggestionsFetchRequested={[Function]}
            renderSectionTitle={[Function]}
            renderSuggestion={[Function]}
            renderSuggestionsContainer={[Function]}
            shouldRenderSuggestions={[Function]}
            suggestions={Array []}
            theme={
              Object {
                "container": "react-autosuggest__container",
                "containerOpen": "react-autosuggest__container--open",
                "input": "react-autosuggest__input",
                "inputFocused": "react-autosuggest__input--focused",
                "inputOpen": "react-autosuggest__input--open",
                "sectionContainer": "react-autosuggest__section-container",
                "sectionContainerFirst": "react-autosuggest__section-container--first",
                "sectionTitle": "react-autosuggest__section-title",
                "suggestion": "react-autosuggest__suggestion",
                "suggestionFirst": "react-autosuggest__suggestion--first",
                "suggestionHighlighted": "react-autosuggest__suggestion--highlighted",
                "suggestionsContainer": "react-autosuggest__suggestions-container",
                "suggestionsContainerOpen": "react-autosuggest__suggestions-container--open",
                "suggestionsList": "react-autosuggest__suggestions-list",
              }
            }
          >
            <Autowhatever
              getSectionItems={[Function]}
              highlightedItemIndex={null}
              highlightedSectionIndex={null}
              id="1"
              inputProps={
                Object {
                  "onBlur": [Function],
                  "onChange": [Function],
                  "onFocus": [Function],
                  "onKeyDown": [Function],
                  "placeholder": "Search Modules",
                  "value": "",
                }
              }
              itemProps={[Function]}
              items={Array []}
              multiSection={true}
              renderInputComponent={[Function]}
              renderItem={[Function]}
              renderItemData={
                Object {
                  "query": "",
                }
              }
              renderItemsContainer={[Function]}
              renderSectionTitle={[Function]}
              theme={
                Object {
                  "container": "react-autosuggest__container",
                  "containerOpen": "react-autosuggest__container--open",
                  "input": "react-autosuggest__input",
                  "inputFocused": "react-autosuggest__input--focused",
                  "inputOpen": "react-autosuggest__input--open",
                  "item": "react-autosuggest__suggestion",
                  "itemFirst": "react-autosuggest__suggestion--first",
                  "itemHighlighted": "react-autosuggest__suggestion--highlighted",
                  "itemsContainer": "react-autosuggest__suggestions-container",
                  "itemsContainerOpen": "react-autosuggest__suggestions-container--open",
                  "itemsList": "react-autosuggest__suggestions-list",
                  "sectionContainer": "react-autosuggest__section-container",
                  "sectionContainerFirst": "react-autosuggest__section-container--first",
                  "sectionTitle": "react-autosuggest__section-title",
                }
              }
            >
              <div
                className="react-autosuggest__container"
              >
                <input
                  aria-activedescendant={null}
                  aria-autocomplete="list"
                  aria-expanded={false}
                  aria-haspopup={false}
                  aria-owns="react-autowhatever-1"
                  autoComplete="off"
                  className="react-autosuggest__input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  placeholder="Search Modules"
                  role="combobox"
                  type="text"
                  value=""
                />
                <div
                  className="react-autosuggest__suggestions-container"
                  id="react-autowhatever-1"
                />
              </div>
            </Autowhatever>
          </Autosuggest>
        </ModuleSearch>
      </div>
      <div
        className="list"
      >
        <ModuleCategory
          collapsed={false}
          isReadOnly={false}
          modules={
            Array [
              <Module
                addModule={[Function]}
                icon="url"
                id={4}
                name="Load from Enigma"
            />,
              <Module
                addModule={[Function]}
                icon="url"
                id={2}
                name="Load from Facebork"
            />,
            ]
          }
          name="Add data"
          setOpenCategory={[Function]}
        >
          <div
            className="card b-l-gray library-card-category-open"
          >
            <div
              className="cat-open"
            >
              <div
                className="first-level d-flex align-items-center"
                onClick={[Function]}
              >
                <div
                  className="cat-container"
                >
                  <div
                    className="icon-sort-down-vl-gray small-sort button-icon-library ml-3 mb-1"
                  />
                  <span
                    className="content-3 t-vl-gray ml-3 ml-cat"
                  >
                    Add data
                  </span>
                </div>
              </div>
              <div>
                <Collapse
                  className=""
                  delay={
                    Object {
                      "hide": 350,
                      "show": 350,
                    }
                  }
                  isOpen={true}
                  onClosed={[Function]}
                  onOpened={[Function]}
                  tag="div"
                >
                  <div
                    className="collapse show"
                    style={Object {}}
                  >
                    <div
                      className="ml-list"
                    >
                      <Module
                        addModule={[Function]}
                        icon="url"
                        id={4}
                        name="Load from Enigma"
                      >
                        <div
                          className="card"
                          style={
                            Object {
                              "border": 0,
                              "borderRadius": 0,
                            }
                          }
                        >
                          <div
                            className=""
                            onClick={[Function]}
                          >
                            <div
                              className="second-level d-flex flex-row align-items-center"
                            >
                              <div
                                className="ml-icon-container"
                              >
                                <div
                                  className="icon-url ml-icon"
                                />
                              </div>
                              <div>
                                <div
                                  className="content-5 ml-module-name"
                                >
                                  Load from Enigma
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </Module>
                      <Module
                        addModule={[Function]}
                        icon="url"
                        id={2}
                        name="Load from Facebork"
                      >
                        <div
                          className="card"
                          style={
                            Object {
                              "border": 0,
                              "borderRadius": 0,
                            }
                          }
                        >
                          <div
                            className=""
                            onClick={[Function]}
                          >
                            <div
                              className="second-level d-flex flex-row align-items-center"
                            >
                              <div
                                className="ml-icon-container"
                              >
                                <div
                                  className="icon-url ml-icon"
                                />
                              </div>
                              <div>
                                <div
                                  className="content-5 ml-module-name"
                                >
                                  Load from Facebork
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </Module>
                    </div>
                  </div>
                </Collapse>
              </div>
            </div>
          </div>
        </ModuleCategory>
        <ModuleCategory
          collapsed={true}
          isReadOnly={false}
          modules={
            Array [
              <Module
                addModule={[Function]}
                icon="chart"
                id={1}
                name="Chartbuilder"
            />,
            ]
          }
          name="Visualize"
          setOpenCategory={[Function]}
        >
          <div
            className="card b-l-gray library-card-category-closed"
          >
            <div
              className="cat-open"
            >
              <div
                className="first-level d-flex align-items-center"
                onClick={[Function]}
              >
                <div
                  className="cat-container"
                >
                  <div
                    className="icon-sort-right-vl-gray button-icon-library ml-3"
                  />
                  <span
                    className="content-3 t-vl-gray ml-3 ml-cat"
                  >
                    Visualize
                  </span>
                </div>
              </div>
              <div>
                <Collapse
                  className=""
                  delay={
                    Object {
                      "hide": 350,
                      "show": 350,
                    }
                  }
                  isOpen={false}
                  onClosed={[Function]}
                  onOpened={[Function]}
                  tag="div"
                >
                  <div
                    className="collapse"
                    style={Object {}}
                  >
                    <div
                      className="ml-list"
                    >
                      <Module
                        addModule={[Function]}
                        icon="chart"
                        id={1}
                        name="Chartbuilder"
                      >
                        <div
                          className="card"
                          style={
                            Object {
                              "border": 0,
                              "borderRadius": 0,
                            }
                          }
                        >
                          <div
                            className=""
                            onClick={[Function]}
                          >
                            <div
                              className="second-level d-flex flex-row align-items-center"
                            >
                              <div
                                className="ml-icon-container"
                              >
                                <div
                                  className="icon-chart ml-icon"
                                />
                              </div>
                              <div>
                                <div
                                  className="content-5 ml-module-name"
                                >
                                  Chartbuilder
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </Module>
                    </div>
                  </div>
                </Collapse>
              </div>
            </div>
          </div>
        </ModuleCategory>
      </div>
      <ImportModuleFromGitHub
        moduleAdded={[Function]}
      >
        <div>
          <div
            className="import-module-button content-3 mb-5 t-vl-gray"
            onClick={[Function]}
          >
            IMPORT FROM GITHUB
          </div>
          ;
          <Modal
            autoFocus={true}
            backdrop={true}
            backdropTransitionTimeout={150}
            className="dialog-window"
            fade={true}
            isOpen={false}
            keyboard={true}
            modalTransitionTimeout={300}
            toggle={[Function]}
            zIndex={1050}
          />
        </div>
      </ImportModuleFromGitHub>
    </div>
  </div>
</ModuleLibrary>
`;
